core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala

```diff
@@ -316,16 +314,30 @@ trait DeltaSourceBase extends Source
    * @param indexedFiles actions iterator from which to generate the DataFrame.
    */
   protected def createDataFrame(indexedFiles: Iterator[IndexedFile]): DataFrame = {
-    val addFilesList = indexedFiles
-        .map(_.getFileAction)
-        .filter(_.isInstanceOf[AddFile])
-        .asInstanceOf[Iterator[AddFile]].toArray
-
-    deltaLog.createDataFrame(
-      readSchemaSnapshotDescriptor,
-      addFilesList,
-      isStreaming = true
-    )
+    // Read AddFiles from different versions in different scans.
+    // This avoids an issue where we might read the same file with different deletion vectors in
+    // the same scan, which we cannot support as long we broadcast a map of DVs for lookup.
+    // This code can be removed once we can pass the DVs into the scan directly together with the
+    // AddFile/PartitionedFile entry.
+    indexedFiles
```

Comment:
This seems to affect every table whether it has DVs or not. Any perf/regression risk here given the union of dataframes?

Is it possible to restrict the new logic for tables with DVs?

Reply:
That's fair. I changed to only do the union thing when there are actually DVs involved in the current file list.